### DISCUSSION TITLE
Communicating with global state singleton

### DISCUSSION THREAD

**Discussion by geekox86** [statelyai/xstate#3440] [Q&A]
Hi again,

I have a UI hierarchy modelled nicely using xstate where each UI component machine is invoked by its parent component machine. I also have query machines inside some of these components.

Now, I need to add a global cache machine that query machines need to communicate with before/after they fetch data over the network. Also, I do not want to manually pass references down the hierarchy to facilitate this communication.

Where should I place my global state machines like the cache in my hierarchy?

What are my options to enable communication between local state machines and global state machines without manually passing around references?

I wish there is a dependency injection kind of feature in xstate.

Thank you.

**Comment by davidkpiano** [comment]
> I wish there is a dependency injection kind of feature in xstate.

What would this look like to you? Pseudocode welcome

**Comment by Andarist** [comment]
It feels like [the receptionist RFC](https://github.com/statelyai/rfcs/pull/5) could make this easier for you.

**Reply by geekox86** [reply]
Yes, much easier!
Is this available for now in v4?
Do I have access to try v5 beta?

**Reply by davidkpiano** [reply]
It is not available yet but we will eventually make a PR for it. Please share your feedback in the RFC before then!

**Comment by zacharyweidenbach** [comment]
If you need a solution for this problem now, I asked basically the same question a month or so ago.

https://github.com/statelyai/xstate/discussions/3345

In my project, there are global running machines and, as I call them, ephemeral UI machines that mirror the lifecycle of the components they model. But the principle is the same. I needed a UI machine to be able to communicate with a global query machine (which also holds the api cache data/errors). The solution that I went with was recommended by Chris Shank. It is essentially an event bus that connects two machines such that they can send and receive events to each other.

Here is a test file that verifies the correctness of this behavior:
1. The Event Bus logic (taken directly from Chris Shank's repo: https://github.com/ChrisShank/xstate-behaviors/blob/b7484ace3a5445f75488c6dd8e4e5d4204692944/src/event-bus/from-event-bus.ts)
https://github.com/zacharyweidenbach/react-native-test-sandbox/blob/893e24f93628637cb512cadfdc869f649a6a91f4/NativeFSM/utils/xstate/EventBus.ts
2. A query machine and UI machine that are created and connected with the same event bus
https://github.com/zacharyweidenbach/react-native-test-sandbox/blob/893e24f93628637cb512cadfdc869f649a6a91f4/NativeFSM/store/utils/queryMachineWithColdStorage/index.integration.spec.ts#L39 . Logic on the UIMachine to return to the `idle` state if the queryMachine emits a RESET event
https://github.com/zacharyweidenbach/react-native-test-sandbox/blob/893e24f93628637cb512cadfdc869f649a6a91f4/NativeFSM/store/utils/queryMachineWithColdStorage/index.integration.spec.ts#L58
4. Test verifying the UI machine returns to the `idle` state when the query machine (connected via event bus) emits a RESET event
https://github.com/zacharyweidenbach/react-native-test-sandbox/blob/893e24f93628637cb512cadfdc869f649a6a91f4/NativeFSM/store/utils/queryMachineWithColdStorage/index.integration.spec.ts#L127

**Reply by geekox86** [reply]
Thank you very much for the approach
